### ISSUE TITLE
w_eddist fails to run

### ISSUE THREAD

**Issue by ajoshpratt** [westpa/west_tools#2]
Errors out with the following when run with the following command (after running w_assign and w_kinetics with two states specified):

w_eddist --istate 0 --fstate 1

Traceback (most recent call last):                                                                                                                                                                         ]
  File "/home/varus/apps/westpa/westpa-current/lib/wwmgr/work_managers/processes.py", line 71, in task_loop
    result = fn(*args, **kwargs)
  File "/home/varus/apps/westpa/westpa-current/lib/west_tools/w_eddist.py", line 83, in _remote_bin_iter
    histnd(dset[:,ipt,:], binbounds, weights, out=iter_hist, binbound_check = False, ignore_out_of_range=ignore_out_of_range)
  File "_fasthist.pyx", line 39, in fasthist._fasthist.histnd (fasthist/_fasthist.c:3640)
  File "_fasthist.pyx", line 108, in fasthist._fasthist.histnd (fasthist/_fasthist.c:2872)
  File "_fasthist.pyx", line 226, in fasthist._fasthist._histnd (fasthist/_fasthist.c:8684)
ValueError: value 0.0 at index 0 out of bin boundaries in dimension 0

Possible change in underlying library; creating more than two states and specifying --istate 1 and --fstate 2 (or some combination thereof) does seem to get around this bug, but it's unclear whether or not it is producing meaningful output in that scenario.  I am unsure if they are two bugs or are related.

**Comment by mczwier** [comment]
This error usually shows up as a result of binning trouble, as in a data point is falling outside of your defined bin region (here, 0.0 doesn't find itself in a bin). Do the bins for this run make sense?

**Comment by ajoshpratt** [comment]
Interestingly enough, the bins seem to be fine for w_assign, w_kinetics and w_kinavg; as well, 0.0 is not a physically attainable progress coordinate in this system, so while I have not hand checked to see whether it exists in main .h5 file, it seems unlikely.  In addition, the bins for this system are as follows, with the RecitilinearBinMapper:

[[0.0, 4.0, 10.00, 1000000]]

**Comment by mczwier** [comment]
I imagine that w_eddist is using the fast histogram routines to bin transition durations, and it has received a zero-length transition duration. That would seem to indicate either that w_kinetics is dumping zero-length transitions and w_eddist is not expecting it to, or that w_eddist is trying to bin some uninitialized data somewhere.

**Comment by mczwier** [comment]
Or, I suppose, that w_eddist has incorrect bins set up for binning event durations.

**Comment by ajoshpratt** [comment]
When the tool was originally being developed, some modifications did need to be made to w_kinetics to output data of the proper type (although, I no longer remember, off the top of my head, what they were, specifically); something about throwing out a little extra data when it was doing the trace.  

I think I recall that change being pushed to the main repo, but it's possible an older version was somehow pushed back up at some point.  I suppose it's worth it to check the edit history for w_kinetics.

**Comment by ajoshpratt** [comment]
It looks like that change is probably still there, since I'm the one that made it and apparently, I'm the last one that edited w_kinetics.  Probably deserves a deeper look.  I can take a look at this, in depth, but not until early March.

**Comment by mczwier** [comment]
Sounds good.

**Comment by ajoshpratt** [comment]
Looks like you're correct about the binning, actually.  By default, without tweaking on the command line, it seems like the tool isn't making sane choices about how to bin over the data (for the bins selected above, it's going from -1 to 0), hence the error.  The logic regarding how to split the default bin parameters should be adjusted or copied from w_pdist (for consistency), tested, and merged.

**Comment by ajoshpratt** [comment]
Part of this seems to come from some slight difference in the histogram binning scheme relative to w_pdist, as w_eddist started life as a fork of it.  We'll see if this is correction is legitimate or needs to be rolled back.

Part of it, however, comes from the perfectly reasonable behaviour of w_kinetics not being able to record any duration data unless a set of interim transition states is specified in the w_assign step.  Specifically, the problem can be nailed down as follows: in the reasonable use case of a two state system with, say, 3 bins, with one bin separating the states, all recorded durations in kinetics.h5 (the default output from w_kinetics) will be of length 0, and w_eddist will fail.

I don't think we should change the behavior of w_kinetics; instead, I think w_eddist should come with documentation advising the end user of proper usage regarding how to assign states for correct usage of w_eddist, as simply failing out with a relatively arcane error is not helpful.  In the event that they do not listen and the tool is faced with a large array of 0s, I think w_eddist should fail gracefully, produce a sane output that integrates to 1 anyway (I mean, what if the tau is long and all the transitions DO happen in one iteration?), and advise the user to check the documentation.

I also went deep into into _kinetics.pyx, which I think means I just volunteered myself for handling some of the docstrings in there (specifically, cpdef find_macrostate_transitions), assuming .pyx files can handle docstrings, as I now understand what the function is doing.

**Comment by mczwier** [comment]
Wait a second...w_kinetics should *not* be giving zero durations for state-to-state transitions in a 3-bin system. Say we have bins A, I, and B (states A and B, and intermediate I). AB and BA durations should be nonzero if at all observed. AI and IA should also be nonzero, but frequently one dt due to fluctuations. The only transitions that should be zero are AA, II, and BB, because there's no clean way to separate round-trips (A-B-A) from lifetimes (just hanging around in A).

**Comment by ajoshpratt** [comment]
I think it would be good if someone else could confirm if they run into the same problem and that it's not just something I'm doing, then.

Oddly, I thought that was consistent with the way w_assign worked: tag a segment not with the state it is currently in, but with the state it was last in (unless it's currently in a macrostate, obviously).  It cannot, therefore, distinguish between when it left a macrostate barrier and entered a new one unless you place an intermediate state between your true states of interest.

**Comment by ajoshpratt** [comment]
The duration length of 0 for a two state system, I mean.

**Comment by mczwier** [comment]
Two states with three bins should work. If it doesn't, then we have bigger problems than w_eddist. w_kinetics was never intended to work without at least one bin separating macrostates, as the definitions of transit time and waiting time get murky (if not break down outright) without at least one bin separating states.

**Comment by ajoshpratt** [comment]
I agree that it should work.  Looking at _kinetics.pyx, it seems the duration field is calculated based on when you last exited the state, which is right before you state flipped.  I was not aware this was not intended behavior, however, as I'm not really sure where this duration value is actually used (given that w_kinavg works fine regardless of what value we have there).

Using the following code snippet with an intermediate state, I get the following result (and w_eddist works):

In [1]: import numpy
In [2]: import h5py
In [3]: a = h5py.File('kintrace.h5')
In [4]: numpy.nonzero(a['durations']['duration'])  
Out[4]: (array([   3,    3,    3, ..., 1999, 1999, 1999]),
 array([2358, 2362, 2365, ..., 2405, 2416, 2432]))

If I remove the intermediate state, however, and use the same code (without adjusting the bins, obviously)
Out[4]: (array([], dtype=int64), array([], dtype=int64))

I can supply state definition and bin files, as well as data, if you need something for replication.

**Comment by mczwier** [comment]
So the problem is that an intermediate bin that does not have an intermediate state explicitly corresponding to it leads to incorrect results?

**Comment by ajoshpratt** [comment]
Seems to be the case.

Perhaps we should open another discussion for this, as it seems unrelated to w_eddist, but unless I'm mistaken and/or reading it incorrectly (quite possible), the routine used by w_kinetics in _kinetics.pyx creates the duration value by subtracting, from the current time, the time the segment last left the 'other' state; however, as w_assign only updates the state assignment once the segment enters a new macrostate, not once you leave it (that would defeat the purpose of state tracking, after all), the segment is considered to have an event duration of 0.

(I so have not mastered github's Markdown):

from _kinetics.pyx:
539                         if iistate != flabel:
540                             t_ed = tm - _last_exits[seg_id,iistate]
541                             durations.append((iistate,flabel,t_ed,_weight, seg_id))

where tm is the current time.  Essentially, without an intermediate state, _last_exits[seg_id,iistate] always equals tm.  Let me know if that's a naive reading of the code, as well as if we should open another discussion.

**Comment by ajoshpratt** [comment]
I should mention that, for an event duration, this seems like desired behavior: we WANT when we last left a state.  However, we obviously don't want to silently fill an h5 with essentially garbage data, either.

**Comment by mczwier** [comment]
This was one of (several) reasons that the "unassigned" state and "unknown" bin were created for the kinetics analysis: so that transitions can have an intermediate state (or at least a not-A-or-B state) even when that state is unspecified. We need to decide if we want to continue that behavior (which is invisible to the user but actually quite confusing to program with), or if we want to specify that a transition event duration analysis needs an explicitly selected intermediate state. In other words: code changes or documentation changes?

**Comment by ajoshpratt** [comment]
I'm not sure we can avoid the necessity of an unknown state.  While in the 3 bin, 2 state example above, the unknown state IS the intermediate state, in a more complex example, I don't know if the unknown tag would necessarily correspond to a true kinetic intermediate.  I'd have to think about it a little more, but if there are bins that exist that are not 'between' the states, I would imagine a need for them to remain 'unknown', I should think.

One possibility is the automatic creation of another state (or another set of states, anyway) that would be invisible to the end user: i to j transition states.  Or, maybe only one is truly necessary?  At any rate, if the user doesn't create the state, we don't necessarily have to be interested in making them formal states (i.e., have bins attached to them), unless a lot of machinery breaks down otherwise: they could just have a number slapped on to them that signifies "I WAS in a macrostate, and now I'm not, anymore".  This could potentially muck with anything that relies on the output from w_assign, such as the non-Markovian post analysis toolkit from @synapticarbors, but I'm not sure, exactly.  We could have it to where the state indices are trivially linked to the state where they came from, so an end user or script could easily determine what the 'true' macrostate the segment was last in was without resorting to reimplementing w_assign (i.e., if we have 2 user defined states, we have 4 virtual states: 2 real, and 2 intermediate, where state 0 and state 1 are the user defined ones, and states 2 and states 3 are the intermediates linked to 0 and 1 or something, and then in this case, I guess, state 4 would be the unknown).  This would absolutely muck with the logic in the post analysis reweighting toolkit, though.  I don't think it would be a huge change there, but.

This isn't the 'purest' solution, though, as it does separate the hard link between bins and states.  Anyway, just a thought.  What functionality relies on the duration output besides w_eddist?  A documentation change would certainly be EASIER, but...

**Comment by mczwier** [comment]
It sounds like we need to clean up (or fix) the existing "unknown state" functionality. All of this should be covered by that, without messing with the post-analysis toolkit.

**Comment by ajoshpratt** [comment]
A branch has been created in westpa tools called ed_kinetics with some initial, "seems to be working" code in place.

In short, I've changed things to do the following:

w_assign (and the underlying routines, such as some of the routines in _assign.pyx) creates a new dataset which is essentially an 'uncolored' macrostate assignment dataset: either you're in a true macrostate, at time t, or you're in an unknown state.  It should be, in essence, a simple matching between the macrostate assignments and bin assignments dataset.  This needs to be confirmed as functioning, but I don't think I made any mistakes.

This seemed like a better option than rigging some fly-by-night counter up that would just do it on the fly inside of a more delicate operation.

In _find_macrostate_transitions in _kinetics.pyx, this dataset is loaded and used to start a second counter for the event durations: whenever the current labeled state and the current macrostate assignment match, the counter is set to the current time.  Once they no longer match (i.e., once they cross the transition barrier threshold into the unknown or a new macrostate), the timer stops being set to the current time and is fixed at whatever it was set to last.  When we transition into a new true macrostate, we calculate the i to j event duration time using this stopwatch.

This can probably be improved more, but that's it for now.
